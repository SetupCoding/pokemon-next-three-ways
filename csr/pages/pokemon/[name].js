import { Col, Container, Row } from "react-bootstrap";

import Head from "next/head";
import axios from "axios";
import { useQuery } from "react-query";
import { useRouter } from "next/router";

const getPokemon = async (key, name) => {
  const { data } = await axios.get(`/api/pokemon?name=${escape(name)}`);
  return data;
};

const PokemonDetails = () => {
  const router = useRouter();
  const { data } = useQuery(["name", router.query.name], getPokemon);
  return (
    <div>
      <Head>
        <title>{(data && data.name.english) || "Pokemon"}</title>
      </Head>
      <Container>
        {data && (
          <>
            <h1>{data.name.english}</h1>
            <Row>
              <Col xs={4}>
                <img
                  src={`/pokemon/${data.name.english
                    .toLowerCase()
                    .replace(" ", "-")}.jpg`}
                  style={{
                    width: "100%",
                  }}
                />
              </Col>
              <Col xs={8}>
                {Object.entries(data.base).map(([key, value]) => (
                  <Row key={key}>
                    <Col xs={2}>{key}</Col>
                    <Col xs={10}>{value}</Col>
                  </Row>
                ))}
              </Col>
            </Row>
          </>
        )}
      </Container>
    </div>
  );
};

export default PokemonDetails;
