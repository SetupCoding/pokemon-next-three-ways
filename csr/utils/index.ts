export {enhanceDataWithImages, addImage} from './enhanceDataWithImages';
export {handleErrors} from './handleErrors';