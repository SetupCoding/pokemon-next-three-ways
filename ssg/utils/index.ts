export {enhanceDataWithImages} from './enhanceDataWithImages';
export {handleErrors} from './handleErrors';