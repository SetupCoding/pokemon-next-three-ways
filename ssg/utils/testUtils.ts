import React from "react";
import renderer from "react-test-renderer";

export const testSnapshot = (component) => {
  return expect(renderer.create(component).toJSON()).toMatchSnapshot();
};
