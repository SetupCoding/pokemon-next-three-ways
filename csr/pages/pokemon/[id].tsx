import { Col, Container, Row } from "react-bootstrap";
import { ErrorComponent, Home, Loading } from "../../components";

import Head from "next/head";
import { Pokemon } from "../../models/pokemon.model";
import { handleErrors } from "../../utils/handleErrors";
import { useQuery } from "react-query";
import { useRouter } from "next/router";

const getPokemon = async (_, id): Promise<Pokemon> => {
  const res = await fetch(`/api/pokemon?id=${escape(id)}`);
  return await handleErrors(res);
};

const PokemonDetails: React.FC = () => {
  const router = useRouter();
  const { data, error, isFetching } = useQuery(
    ["id", router.query.id],
    getPokemon,
    { retry: false, enabled: !!router.query.id }
  );
  return (
    <div>
      <Head>
        <title>{(data && data.name.english) || "Pokemon"}</title>
        <meta
          name="description"
          content={`Base stats of ${data?.name?.english}`}
          key="description"
        />
      </Head>
      <Home />
      <Container>
        {data && (
          <>
            <h1>{data.name.english}</h1>
            <Row>
              <Col xs={4}>
                <img
                  src={`/pokemon/${data.name.english
                    .toLowerCase()
                    .replace(" ", "-")}.jpg`}
                  alt={`Image of ${data.name.english}`}
                  style={{
                    width: "100%",
                  }}
                />
              </Col>
              <Col xs={8}>
                {Object.entries(data.base).map(([key, value]) => (
                  <Row key={key}>
                    <Col xs={2}>{key}</Col>
                    <Col xs={10}>{value}</Col>
                  </Row>
                ))}
              </Col>
            </Row>
          </>
        )}
      </Container>

      <Loading isLoading={isFetching} />
      <ErrorComponent error={error} />
    </div>
  );
};

export default PokemonDetails;
