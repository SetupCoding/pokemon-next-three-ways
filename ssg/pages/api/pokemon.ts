import { NextApiRequest, NextApiResponse } from "next";

import pokemon from "../../pokemon.json";

export default (req: NextApiRequest, res: NextApiResponse) => {
  const idToFind = req.query.id as string;

  if (!idToFind) {
    res.statusCode = 400;
    res.end("Must have a name");
  } else {
    const found = pokemon.find(({ id }) => id === +idToFind);
    if (!found) {
      res.statusCode = 404;
      res.end(
        JSON.stringify({
          error: {
            status: 404,
            error: 'Not Found',
            message: `Pokemon with id ${req.query.id} not found`,
          },
        })
      );
    } else {
      res.statusCode = 200;
      res.setHeader("Content-Type", "application/json");
      res.end(JSON.stringify(found));
    }
  }
};
