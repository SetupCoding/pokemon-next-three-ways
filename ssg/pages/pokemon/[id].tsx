import { Col, Container, Row } from "react-bootstrap";
import { GetStaticPaths, GetStaticProps } from "next";

import Head from "next/head";
import { Home } from "../../components";
import Image from "next/image";
import { Pokemon } from "../../models/pokemon.model";
import React from "react";
import c from "../../components/pokemon-card.module.css";
import pokemon from "../../pokemon.json";

/**
 * Runing at build time
 * */
export const getStaticPaths: GetStaticPaths = async () => {
  return {
    paths: pokemon.map(({ id }) => ({
      params: {
        id: id.toString(),
      },
    })),
    fallback: false,
  };
};

export const getStaticProps: GetStaticProps = async (context) => {
  return {
    props: {
      data: pokemon.find(({ id }) => id === +context.params.id),
    },
  };
};

// export const getServerSideProps: GetServerSideProps = async (context) => {
//   const res = await fetch(`${API_URL}/pokemon?id=${context.params.id}`);
//   const errorCode = res.ok ? false : (res as HttpError).status;
//   const data = await res.json();

//   return {
//     props: {
//       data: data,
//       error: errorCode ? data : null,
//     },
//   };
// };
type Props = {
  data: Pokemon;
};
const PokemonDetails: React.FC<Props> = ({ data }) => {
  // const router = useRouter();
  // const { data, error, isFetching } = useQuery(
  //   ["id", router.query.id],
  //   getPokemon,
  //   { retry: false, enabled: !!router.query.id }
  // );
  return (
    <div>
      <Head>
        <title>{(data && data.name.english) || "Pokemon"}</title>
        <meta
          name="description"
          content={`Base stats of ${data?.name?.english}`}
          key="description"
        />
      </Head>
      <Home />
      <Container>
        {data && (
          <>
            <h1>{data.name.english}</h1>
            <Row>
              <Col xs={4} className={c.detailImage}>
                <Image
                  layout="fill"
                  src={`/pokemon/${data.name.english
                    .toLowerCase()
                    .replace(" ", "-")}.jpg`}
                  alt={`Image of ${data.name.english}`}
                  objectFit="contain"
                />
              </Col>
              <Col xs={8}>
                {Object.entries(data.base).map(([key, value]) => (
                  <Row key={key}>
                    <Col xs={2}>{key}</Col>
                    <Col xs={10}>{value}</Col>
                  </Row>
                ))}
              </Col>
            </Row>
          </>
        )}
      </Container>

      {/* <Loading isLoading={isFetching} /> */}
      {/* <ErrorComponent error={error} /> */}
    </div>
  );
};

export default PokemonDetails;
