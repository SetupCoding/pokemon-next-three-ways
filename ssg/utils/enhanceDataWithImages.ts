import { HOST_URL } from "../constants/const";
import { Pokemon } from "../models/pokemon.model";

export const addImage = (pokemon: Pokemon) => ({
  ...pokemon,
  image: `/pokemon/${pokemon.name.english.toLowerCase().replace(" ", "-")}.jpg`,
});

export const enhanceDataWithImages = (
  data: Omit<Pokemon, "image">[]
): Pokemon[] => data.map((pokemon) => addImage(pokemon));
