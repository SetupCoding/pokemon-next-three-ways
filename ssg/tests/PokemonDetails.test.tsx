import { Pokemon } from "../models/pokemon.model";
import PokemonDetails from "../pages/pokemon/[id]";
import React from "react";
import { testSnapshot } from "../utils/testUtils";

describe("<PokemonDetails />", () => {
  it("renders the PokemonDetails component correctly", () => {
    const data: Pokemon = {
      id: 1,
      name: {
        english: "Bulbasaur",
        japanese: "フシギダネ",
        chinese: "妙蛙种子",
        french: "Bulbizarre",
      },
      type: ["Grass", "Poison"],
      base: {
        HP: 45,
        Attack: 49,
        Defense: 49,
        "Sp. Attack": 65,
        "Sp. Defense": 65,
        Speed: 45,
      },
    };
    testSnapshot(<PokemonDetails data={data} />);
  });
});
