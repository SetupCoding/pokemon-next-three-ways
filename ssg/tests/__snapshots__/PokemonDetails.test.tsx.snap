// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonDetails /> renders the PokemonDetails component correctly 1`] = `
<div>
  <a
    href="/"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    &lt; Home
  </a>
  <div
    className="container"
  >
    <h1>
      Bulbasaur
    </h1>
    <div
      className="row"
    >
      <div
        className="detailImage col-4"
      >
        <img
          alt="Image of Bulbasaur"
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </div>
      <div
        className="col-8"
      >
        <div
          className="row"
        >
          <div
            className="col-2"
          >
            HP
          </div>
          <div
            className="col-10"
          >
            45
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-2"
          >
            Attack
          </div>
          <div
            className="col-10"
          >
            49
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-2"
          >
            Defense
          </div>
          <div
            className="col-10"
          >
            49
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-2"
          >
            Sp.Attack
          </div>
          <div
            className="col-10"
          >
            65
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-2"
          >
            Sp.Defense
          </div>
          <div
            className="col-10"
          >
            65
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-2"
          >
            Speed
          </div>
          <div
            className="col-10"
          >
            45
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
